cmds/fspeed.py

```diff
@@ -0,0 +1,141 @@
+import math
+import json
+from dataclasses import dataclass, field
+
+from discord import Message
+
+from core.keyboard import Layout, Position
+from util import authors, corpora, memory, parser
+
+FINGERS = ('LP', 'LR', 'LM', 'LI', 'RI', 'RM', 'RR', 'RP')
+
+
+def exec(message: Message):
+    id = message.author.id
+    kwargs = parser.get_kwargs(message, str,
+                               stagger=bool, lateral=str, sfb=str, dsfb=str, sfs=str, key_travel=str,
+                               kps=list, weights=list)
+    ll_name: str = kwargs['args']
+
+    if not ll_name:
+        return ('```\n'
+                'fspeed [layout name] [--kwargs]\n'
+                'see the finger speed of a layout\n'
+                'Options:\n'
+                '--stagger\n'
+                '--[lateral, sfb, dsfb, key_travel] [num]\n'
+                '--kps [num...]\n'
+                '```')
+
+    kwargs['dsfb'] = kwargs['dsfb'] or kwargs['sfs']
+    kwargs['kps'] = kwargs['kps'] or kwargs['weights']
+    try:
+        fspeed_params = {}
+        if kwargs['stagger']:
+            fspeed_params['stagger'] = True
+        if kps := kwargs['kps']:
+            fspeed_params['kps'] = parse_kps(kps)
+        for param in ('lateral', 'dsfb', 'key_travel'):
+            if not (num_str := kwargs[param]):
+                continue
+            fspeed_params[param] = parse_number(num_str)
+    except ValueError as err:
+        return str(err)
+
+    ll = memory.find(ll_name)
+    fspeed = FSpeed(**fspeed_params)
+    unweighted_speeds, weighted_speeds = fspeed.fingerspeed(ll, id)
+
+    with open('likes.json', 'r') as f:
+        likes: dict[str, list[int]] = json.load(f)
+
+    like_count = len(likes.get(ll.name, ()))
+    like_string = 'like' if likes == 1 else 'likes'
+
+    res = ['```',
+           f'{ll.name} ({authors.get_name(ll.user)}) ({like_count} {like_string})',
+           f'{corpora.get_corpus(id).upper()}:',
+           'Unweighted Speed']
+    for finger in FINGERS:
+        res.append(f'    {finger}: {unweighted_speeds[finger]:.3f}')
+    res.extend(('', 'Weighted Speed'))
+    for finger in FINGERS:
+        res.append(f'    {finger}: {weighted_speeds[finger]:.3f}')
+    res.append('```')
+    return '\n'.join(res)
+
+
+@dataclass
+class FSpeed:
+    stagger: bool = False
+    lateral: float = 1.4
+    sfb: float = 1.0
+    dsfb: float = 0.5
+    key_travel: float = 0.01
+    kps: dict[str, float] = field(default_factory=lambda:
+                                  dict(zip(FINGERS,
+                                       (1.5, 3.6, 4.8, 5.5, 5.5, 4.8, 3.6, 1.5))))
+
+    def fingerspeed(self, ll: Layout, id: int) -> tuple[dict[str, float], dict[str, float]]:

```

Comment:
Real fspeed relies on sf2+s so here is a bit of future-proofing even though tetragrams won't be added unless someone finds the raw data for the other corpora that I don't have

```
def fingerspeed(self, ll: Layout, id: int) -> tuple[dict[str, float], dict[str, float]]:
        ngrams: list[tuple[dict[str, float], float, dict[str, float]]] = []
        speeds = {finger: 0.0 for finger in FINGERS}

        weights = [self.sfb, self.dsfb]
        while len(weights) < len(corpora.NGRAMS) - 1: #PLACEHOLDER SINCE I DON'T KNOW 
            weights.append(weights[-1] / 4)           #HOW LARGER NGRAMS ARE WEIGHTED
        
        for length in range(2, len(corpora.NGRAMS) + 1):
            ngrams.append((corpora.ngrams(length, id=id), weights[length - 2], speeds))
        
        for ngram, weight, speeds_dict in ngrams:
            ngram_total = sum(ngram.values())
            for gram, freq in ngram.items():
                p1 = ll.keys.get(gram[0].lower(), None)
                p2 = ll.keys.get(gram[-1].lower(), None)
                if not p1 or not p2 or p1.finger != p2.finger:
                    continue
                dist = self.two_key_dist(p1, p2, True) + 2 * self.key_travel
                speeds_dict[p1.finger] += freq * dist * weight / ngram_total

        unweighted_speeds = {finger: speed * 800 for finger, speed in speeds.items()}
        weighted_speeds = {finger: speed / self.kps[finger] for finger, speed in unweighted_speeds.items()}
        return unweighted_speeds, weighted_speeds
```

You should probably consult semi to make larger ngrams more accurate, but this should return the same values with bigrams and trigrams while making it more dynamic.